Support sending challenges with a link

karamnaser opened the pull request:
ther was some issue with code formation had to remove my previous PR
videos link related to issue:
https://streamable.com/l1p99k
https://streamable.com/hn0x32
https://streamable.com/t641dw
in this PR :
[ x ]fixed my code text formation
[ x ]created  popuplisener and added to client in this.socket.onmessege function
[ x ]changed promise to call back in onteamselect
[ x ]changed alert to popups 
[ x ]used Storage.prefs instead of localstoreg 
[ x ]nested all my function inside ManeManuRoom class

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -81,6 +81,18 @@
+				}
+				});
+				self.createLoginFormObserver(self);
+				window.Popuplesiner=self.Popuplesiner;
```
?_?
What's this?

mia-pi-git replied:
Ah. It would be preferred to just use `app.on` and `app.trigger` instead.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +126,131 @@
 				);
 			}
 		},
+		createLoginFormObserver:function(self){
+			const config = { attributes: true, childList: true, subtree: true };
+			const observer = new MutationObserver(function(){
+			if(window.app.user.get("named") && window.Storage.prefs("page")=="challenge"){
+				window.Storage.prefs("page","testclient");
+				if($('.challenge-container').length==0){
+					//user log in event trigerd when user log in and make user pick a team
+					$('body').on("login_event",function(){
+					self.onUserLogIn(self);
+				})
+				$('body').trigger("login_event");
+				}
+			}
+			});
+			observer.observe(document.activeElement, config);
+			},
+		onUserLogIn:function(self){
```
Run eslint please.

karamnaser replied:
ther was issue when runing eslint it adds taps and spaces , and it also change strings to template strings

mia-pi-git replied:
> ther was issue when runing eslint it adds taps and spaces

Yes, that's proper behavior. It's supposed to do that to help format your code properly.

> t also change strings to template strings

I'd advise removing them after running eslint manually.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +126,131 @@
+				$('.challenge-container').empty();
+			})
+		},
+		onTeamSelect:function(challengerName,self){
+			$('#accept-challenge').html("processing request...");
+				if(window.Storage.prefs("team")!="random"){
+					app.sendTeam(window.Storage.teams[$('#select-team .select.teamselect').val()]);
+					app.send('/vtm ' + $('.select.formatselect').val());
+					self.validateChallengeAndSend(self,{userId:challengerName,massage:window.Storage.prefs("errormassege")});
+				}
+				else{
+					self.validateChallengeAndSend(self,{userId:challengerName,massage:"Your team is valid"});
+				}
+		},
```
Another example block that needs eslint.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +126,131 @@
+			$('.pmbox').append(`<div class=challenge-container>${buf}</div>`);
+			$('#generate-link').click(function(){
+			$('#generate-link').html("generating...");
+			if($('.select.teamselect').val()!="random"){
+				app.sendTeam(Storage.teams[$('.select.teamselect').val()]);
+				app.send('/vtm ' + $('.select.formatselect').val());
+			}else if($('.select.teamselect').val()==="random"){
+				app.send("/utm null");
+			}
+			setTimeout(function(){
+			if($('.select.teamselect').val() && (window.Storage.prefs("errormassege").includes("Your team is valid") || $('.select.teamselect').val()==="random")){
+			self.generateLink({userId:window.app.user.get("name"),
+			format:$('.select.formatselect').val()
+			,teamIndex:$('.select.teamselect').val()});
+			$('#generate-link').html("generate link");
+			}
+			else{
+				$('.link').remove();
+				$('#generate-link').html("generate link");
+				window.app.addPopupMessage(window.Storage.prefs("errormassege"));
+				
+			}
+			},2000);
+			});
```
Another example block that needs eslint.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +126,131 @@
+						window.app.addPopupMessage(data.massage);
+					}
+		},
+		generateLink:function(challengeData){
+			if(!$('.link').length){
+			let link="https://play.pokemonshowdown.com?challenge-"+challengeData.userId+"-"+challengeData.format+"-"+challengeData.teamIndex;
+			$('.link-generatore').append("<p class=link>"+link+"</p>");
+			}
+			else{
+				let link="https://play.pokemonshowdown.com?challenge-"+challengeData.userId+"-"+challengeData.format+"-"+challengeData.teamIndex;
+				$('.link').html("<p class=link>"+link+"</p>")
+			}
+		},
```
Another example block that needs eslint.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +126,131 @@
+			buf += '<p><label class="checkbox"><input type="checkbox" name="private" ' + (Storage.prefs('disallowspectators') ? 'checked' : '') + ' /> <abbr title="You can still invite spectators by giving them the URL or using the /invite command">Don\'t allow spectators</abbr></label></p>';
+			buf +='<button type="button" id=generate-link>generate link</button>';
+			buf +='<button type="button" id=close-challenge>close</button></form></div>'
+			$('.pmbox').append(`<div class=challenge-container>${buf}</div>`);
```
```suggestion
			$('.pmbox').append('<div class=challenge-container>' + buf + '</div>');
```

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +126,131 @@
+				$('.challenge-container').empty();
+			})
+		},
+		sendChallenge:function(challengeDataObject){
```
Why not just use `userid, format` args here?

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +126,131 @@
+		sendChallenge:function(challengeDataObject){
+			app.send('/challenge ' + challengeDataObject.userId + ', ' + challengeDataObject.format);
+		},
+		Popuplesiner:function(msg){
+			if(msg.includes("popup")){
+				window.Storage.prefs("errormassege");
+			}
+		},
```
What's this? It doesn't seem to do anything.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -1487,5 +1623,21 @@
 	});
-
+	window.addEventListener("load",function(){
+		if(window.location.search.includes("challenge")){
+			window.Storage.prefs("page","challenge");
+			setTimeout(function(){
+			//changing url without refreshing page, ther was issue of losing data when refreshing page 
+			window.history.replaceState(null,"",'https://play.pokemonshowdown.com/');
+			//log in user before sending challenge
+			new window.UserPopup({name:''}).login();
+			},1000);
+			data=window.location.search.split("-");
+			userId=data[1];
+			format=data[2];
+			window.Storage.prefs("challenger",userId);
```
Another example block that needs eslint.

mia-pi-git commented:
In the future, please just update your old PR instead of opening a new one.

For the use of those looking - [the previous PR is here](https://github.com/smogon/pokemon-showdown-client/pull/1969#discussion_r864520541).

KrisXV commented on js/client-mainmenu.js:
```diff
@@ -54,8 +54,8 @@
-			buf += '<p><button class="button mainmenu5 onlineonly disabled" name="finduser">Find a user</button></p></div>';
-
+			buf += '<p><button class="button mainmenu5 onlineonly disabled" name="finduser">Find a user</button></p>';
+			buf += "<p class=link-generatore><button id='challeng-via-link' class='button mainmenu5 onlineonly'>send challeng via link</button></p></div>";
```
```suggestion
			buf += '<p class="link-generator"><button id="challenge-via-link" class="button mainmenu5 onlineonly">Send challenge via URL</button></p></div>';
```

karamnaser commented:
I attached some vedios about the feature can you check if   the feature work as intended in game?

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +132,140 @@
+		},
+		generateLink:function(challengeData){
+			if(!$('.link').length){
+			let link="https://"+window.Config.routes.client+"?challenge-"+challengeData.userId+"-"+challengeData.format+"-"+challengeData.teamIndex;
```
Why declare this twice? You never change it, so why not just declare it above the if statement?

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -1487,5 +1638,21 @@
+			window.Storage.prefs("page","challenge");
+			setTimeout(function(){
+			//changing url without refreshing page, ther was issue of losing data when refreshing page 
+			window.history.replaceState(null,"","https://"+window.Config.routes.client+"/");
```
Don't hardcode `https:`, use `location.protocol`.

mia-pi-git commented on js/client.js:
```diff
@@ -913,6 +913,8 @@ function toId() {
 		 * Receive from sim server
 		 */
 		receive: function (data) {
+			//trigger popup lisener when messege is received from server and save to storeg.pref obj  
+			app.trigger("lisenToPopUpMessege",{msg:data});
```
This is not necessarily accurate - this can be things other than a popup. That seems unintentional. I'd wait to trigger this until the `switch` statement later down that checks message types. Just call it in the `popup` case.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +132,140 @@
+			}
+		},
+		generateChallengeForm:function(self){
+			var buf = '<div class="challenge"><form class="battleform"><p>Challenge link generatore</p>';
```
```suggestion
			var buf = '<div class="challenge"><form class="battleform"><p>Challenge link generator</p>';
```

mia-pi-git reviewed (changes requested):
Please run eslint.

karamnaser commented:
done the require changes , made some changes on how challenge is triggered via link

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -81,6 +81,17 @@
 			Storage.whenPrefsLoaded(function () {
+			//user log in observer that trigger challenge form when use log in
+				self.createLoginFormObserver(self);
+				$('#challeng-via-link').click(function () {
```
```suggestion
				$('#challenge-via-link').click(function () {
```

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -81,6 +81,17 @@
+							self.generateChallengeForm(self);
+						}
+					} else {
+						window.app.addPopupMessage("you need to log in first to generate link");
```
```suggestion
						app.addPopupMessage("You need to be logged in to generate a challenge link.");
```

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +125,68 @@
+					window.Storage.prefs("page", "testclient");
+					if ($('.challenge-container').length == 0) {
+						//user log in event trigerd when user log in and make user pick a team
+						$('body').on("login", function () {
+							self.onUserLogIn(self);
+						});
+						$('body').trigger("login");
```
Why do you add a listener and then immediately trigger it? Why not just call the function directly?

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +125,68 @@
+		generateLink:function (userId, format) {
+			if (!$('.link').length) {
+				let link = window.location.protocol+"//" + window.Config.routes.client + "?challenge-" + userId + "-" + format;
+				$('.link-generatore').append("<p class=link>" + link + "</p>");
```
```suggestion
				$('.link-generator').append("<p class=link>" + link + "</p>");
```

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +125,68 @@
+		generateChallengeForm:function (self) {
+			var buf = '<div class="challenge"><form class="battleform"><p>Challenge link generatore</p>';
+			buf += '<p><label class="label">Format:</label>' + self.renderFormats("gen8randombattle") + '</p>';
+			buf += '<button type="button" id=generate-link>generate link</button>';
```
```suggestion
			buf += '<button type="button" id=generate-link>Generate link</button>';
```

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +125,68 @@
+			buf += '<button type="button" id=close-challenge>close</button></form></div>';
+			$('.pmbox').append("<div class=challenge-container>" + buf + "</div>");
+			$('#generate-link').click(function () {
+				$('#generate-link').html("generating...");
```
```suggestion
				$('#generate-link').html("Generating...");
```

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +125,68 @@
+			var buf = '<div class="challenge"><form class="battleform"><p>Challenge link generatore</p>';
+			buf += '<p><label class="label">Format:</label>' + self.renderFormats("gen8randombattle") + '</p>';
+			buf += '<button type="button" id=generate-link>generate link</button>';
+			buf += '<button type="button" id=close-challenge>close</button></form></div>';
```
```suggestion
			buf += '<button type="button" id=close-challenge>close</button></form></div>';
```
Element ids are bad practice in this project - you should be using the element's name and then checking which element it is in jQuery event listeners that are set on it.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +125,68 @@
+			const config = {attributes: true, childList: true, subtree: true};
+			const observer = new MutationObserver(function () {
+				if (window.app.user.get("named") && window.Storage.prefs("page") == "challenge") {
+					window.Storage.prefs("page", "testclient");
```
?

karamnaser replied:
the code check when the user log in from the link and then activate the user login event

mia-pi-git replied:
I'm asking what's with the `Storage.prefs('page')` call. It doesn't appear to be needed? If it is - why is it hardcoded?

karamnaser replied:
there was an issue that trigger challenge after user log in from normal game link and not from the challenge link
so i needed to check if the user logged in after inserting challenge link and not normal link

mia-pi-git replied:
Could you not just check the URL for that information? The `challenge-[]-[]` string should still be there and would be useful for that.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +125,68 @@
 		},
+		createLoginFormObserver:function (self) {
+			const config = {attributes: true, childList: true, subtree: true};
+			const observer = new MutationObserver(function () {
```
Why is this necessary? Can you not just use jQuery?

mia-pi-git replied:
Yeah, but I'm asking if you can't just use jQuery to watch the DOM instead? Or why not set a listener trigger in the code that handles completing logins?

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -54,8 +54,8 @@
-			buf += '<p><button class="button mainmenu5 onlineonly disabled" name="finduser">Find a user</button></p></div>';
-
+			buf += '<p><button class="button mainmenu5 onlineonly disabled" name="finduser">Find a user</button></p>';
+			buf += '<p class=link-generatore><button id="challeng-via-link" class="button mainmenu5 onlineonly">send challenge via link</button></p></div>';
```
```suggestion
			buf += '<p class=link-generator><button class="button mainmenu5 onlineonly">Create challenge link</button></p></div>';
```
Button ids not preferred.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +125,68 @@
+			} else {
+				let link = window.location.protocol+"//" + window.Config.routes.client + "?challenge-" + userId + "-" + format;
+				$('.link').html("<p class=link>" + link + "</p>");
+				$('#generate-link').html("generate link");
```
```suggestion
				$('#generate-link').html("generate link");
```
This should be using element names, not ids.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -114,7 +125,68 @@
+			if (!$('.link').length) {
+				let link = window.location.protocol+"//" + window.Config.routes.client + "?challenge-" + userId + "-" + format;
+				$('.link-generatore').append("<p class=link>" + link + "</p>");
+				$('#generate-link').html("generate link");
```
```suggestion
				$('#generate-link').html("generate link");
```
This should be using element names, not ids.

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -1487,5 +1559,20 @@
+			window.Storage.prefs("page", "challenge");
+			setTimeout(function () {
+			//changing url without refreshing page, ther was issue of losing data when refreshing page
+				window.history.replaceState(null, "", window.location.protocol+"//" + window.Config.routes.client+"/");
```
```suggestion
				window.history.replaceState(null, "", window.location.protocol + "//" + window.Config.routes.client + "/");
```

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -1487,5 +1559,20 @@
+		if (window.location.search.includes("challenge")) {
+			window.Storage.prefs("page", "challenge");
+			setTimeout(function () {
+			//changing url without refreshing page, ther was issue of losing data when refreshing page
```
```suggestion
				// changing url without refreshing page, ther was issue of losing data when refreshing page
```

mia-pi-git commented on js/client-mainmenu.js:
```diff
@@ -1487,5 +1559,20 @@
+			let data = window.location.search.split("-");
+			let userId = data[1];
+			let format = data[2];
+			window.Storage.prefs("challenger", userId);
```
Why is Storage.prefs necessary? This data shouldn't be needed across several sessions.

karamnaser commented:
changed observer to app.once(login),removed id from html component ,removed prefstoreg instance and replaced them with 
obj in global scop,fixed spilling errors

AnnikaCodes commented on js/client-mainmenu.js:
```diff
@@ -1,4 +1,9 @@
 (function ($) {
+	var challengeData = {
+	   userId :null,
+	   format : null,
+	   isChallengeTriggerd:false
+	};
```
```suggestion
	var challengeData = {
	   userId: null,
	   format: null,
	   isChallengeTriggered: false,
	};
```

AnnikaCodes commented on js/client-mainmenu.js:
```diff
@@ -54,8 +59,8 @@
-			buf += '<p><button class="button mainmenu5 onlineonly disabled" name="finduser">Find a user</button></p></div>';
-
+			buf += '<p><button class="button mainmenu5 onlineonly disabled" name="finduser">Find a user</button></p>';
+			buf += '<p class=link-generator><button class="challenge-via-link button mainmenu5 onlineonly">send challenge via link</button></p></div>';
```
```suggestion
			buf += '<p class=link-generator><button class="challenge-via-link button mainmenu5 onlineonly">Send challenge via link</button></p></div>';
```

AnnikaCodes commented on js/client-mainmenu.js:
```diff
@@ -81,6 +86,26 @@
 			Storage.whenPrefsLoaded(function () {
+			//user log in event that trigger challenge form when user log in
+				app.once('init:choosename', function () {
+					if (challengeData.isChallengeTriggerd) {
+						challengeData.isChallengeTriggerd = false;
```
```suggestion
					if (challengeData.isChallengeTriggered) {
						challengeData.isChallengeTriggered = false;
```

AnnikaCodes commented on js/client-mainmenu.js:
```diff
@@ -1487,5 +1557,18 @@
+			challengeData.isChallengeTriggerd = true;
+			setTimeout(function () {
+			//changing url without refreshing page, ther was issue of losing data when refreshing page
+				window.history.replaceState(null,"", window.location.protocol + "//" + window.Config.routes.client + "/");
```
```suggestion
				window.history.replaceState(null, "", window.location.protocol + "//" + window.Config.routes.client + "/");
```

AnnikaCodes reviewed (changes requested):
Have you tested this locally?

monsanto commented:
Thanks for the PR but it doesn't seem like you have addressed further feedback so I will assume that you are no longer interested.
